### DISCUSSION TITLE
Possible to generate training data from proofread MusicXML file(s)?

### DISCUSSION THREAD

**Discussion by hallvors** [Audiveris/audiveris#468] [Q&A] [answered]
I am experimenting with Audiveris to convert PDF scores from various sources into MusicXML files. I have several PDFs for different voices in the same piece (probably exported as PDF from some note setting program, this is not scanned images), and used the Docker-based Audiveris to convert all of them - with somewhat mixed results. There are some half-empty files, some null pointer exceptions in the log, but some files also converted pretty well with minor errors. 

Now, having proofread and corrected one file it would be sweet if I could somehow turn that human-verified file into training data and re-recognise the rest of the files with possibly better accuracy.. Are there any command line arguments that amount to "generate and save training data by comparing this source file (or some given processing data) with this desired output"? Would this be an interesting approach to improve accuracy, or is it not likely to make much of a difference?

(Also, let me say that this is a very interesting project and I'm happy to see there's active development happening :). Thank you, maintainers and contributors, for your efforts!)

**Comment by hbitteur** [comment, marked as answer]
@hallvors Thank you for your kind words and your valuable suggestions.

If you have not done it yet, I would recommend that you have a look at:
- [Audiveris omr-dataset-tools project](https://github.com/Audiveris/omr-dataset-tools) (about 4 years old)
- [Audiveris Wiki](https://github.com/Audiveris/audiveris/wiki) and especially its chapter on 6.0 prototype and 6.0 ideas (about 2 years old). You will be introduced to the notions of page classifier and patch classifier.

On top of this, I spent several months last year experimenting the idea of a "head-classifier" (see [head-classifier branch](https://github.com/Audiveris/audiveris/tree/head-classifier) on github).
The head classifier was meant to be a down-sized version of the patch classifier, limited to the recognition of heads in the vicinity of stem seeds, as an alternative to template-matching techniques.
In short, the results were disappointing, the head classifier did not perform up to our expectations.

My personal feeling is that the root cause for this failure was the lack of "true" training material. We have thousands of score images from ZHAW university, but they are just synthetic images. This means too high symbols quality (although we could cope with this in some way) and, more importantly, always the same music layout (because always performed by the same music renderer, the one from Lilypond).

I think what we need is a dataset of real printed music in its wide diversity, together with its related ground truth, in the form of validated .omr files. This can come only from the community of Audiveris users that would agree to upload their .omr files to a common repository.
This does not imply to share these files with the whole world but only with the team of Audiveris project members.

We would need to find a free hosting solution for all these .omr contributions. **Meanwhile, please never throw away your validated .omr files, they represent a very valuable training material!**

**Reply by hallvors** [reply]
Thanks for responding, @hbitteur ! I did not know about the omr-dataset-tools project and I will certainly have a look. I'm relatively new to the whole scenery of OMR and don't know the state of affairs :). 

I'm volunteering for a children's marching band and would very much like to find a way to digitize their old sheet music collections so the conductors can make more use of those.  That's probably thousands of pages with very varying style and quality. It won't get scanned over night, but if you need a range of source data to train things on I suppose I can supply materials.. The proofing will take some time though. I see from the manual that you have some UI for proofing, but as I have just run Audiveris as a command-line tool I haven't actually seen the UI yet. I did the proofing by comparing the rendering in MuseScore to the original, and editing the MusicXML file directly to fix things. I suppose this approach did not generate any "validated .omr files"?

Otherwise, I'm not fluent in the complexities of music engraving, but I have actually used Lilypond for small projects. 

The 6th version and the neural training sounds exciting.

Some random questions, hopefully not too obvious from Wikis and manuals:

1. In the context of version 6, the wiki mentions running a Python service with tensorflow locally - I assume this means I'd train (and benefit from training of) only my own local Audiveris, so to speak? There are no plans for a shared neural network in the cloud which might get trained a lot faster if used by several humans?
2. If you use sheet music from the MuseScore website (such a project was mentioned in a wiki page), I suppose MusicXML will also often be available (right?) - this is analoguous to my situation where I'd like to be able to tell Audiveris "this is the verified outcome I wanted" - would it be useful for the training to have something generating "validated .omr files" from MusicXML?
3. If I at some point get started scanning & proofing music, how would you like training materials contributed? Pull requests against the omr-dataset-tools project? The scanned images + the .omr files after using Audiveris UI to proof stuff is of interest?

**Reply by hallvors** [reply]
I forgot one last question: are bug reports about null pointer exceptions from processing those PDFs I was experimenting with interesting, or is the main development focus going into the new 6.x stuff?

**Reply by gorgobacka** [reply]
> We would need to find a free hosting solution for all these .omr contributions. **Meanwhile, please never throw away your validated .omr files, they represent a very valuable training material!**

I never thought about the importance of the .omr files. I will keep it from now on and organized. If there is a way to contribute those at a later timepoint, I will do it.

**Reply by hbitteur** [reply]
> I forgot one last question: are bug reports about null pointer exceptions from processing those PDFs I was experimenting with interesting, or is the main development focus going into the new 6.x stuff?

6.0 prototype was meant to be a proof of concept: integrating the page classifier from ZHAW into Audiveris architecture and observing its behavior on a few typical input scores.  The page classifier could be seen as a rather global (top down) approach while current Audiveris engine was and is still a rather analytical (bottom up) approach. We didn't have the specific hardware resources needed to retrain the page classifier, and were not satisfied by the early results of the patch classifier either. We tried to develop a head classifier but did not succeed, like I said in my previous answer.

We still have in mind the idea that significant improvement of the OMR engine will require the use of a large representative and validated dataset. And to this end, we decided to work on the UI tools to allow the end user to easily validate / correct the raw engine outputs.

The main development focus today in on 5.x series. 5.0 introduced the notion of .omr project file (in 4.x everything had to be done in one shot in memory!), 5.1 provided a basic UI, 5.2 will provide extensive UI features and a better support of poor-quality scores by the OMR engine. To give you a concrete idea, our target is to be able to process most of IMSLP scores.

The 5.2 release is around the corner (about one month). Meanwhile, if you encounter null pointer exceptions or other problems, do not hesitate to file an issue. Make sure however to use the very latest version by pulling and building from the "development" branch (its name is displayed as 5.2.1-alpha, rather than the now old 5.1 release).

**Reply by hbitteur** [reply]
> 1. In the context of version 6, the wiki mentions running a Python service with tensorflow locally - I assume this means I'd train (and benefit from training of) only my own local Audiveris, so to speak? There are no plans for a shared neural network in the cloud which might get trained a lot faster if used by several humans?

There is a misunderstanding here. The Python service referred to is just meant to _use_ the network locally, in _inference mode_ if you prefer: it takes the whole input image, runs the page classifier on it (a matter of several minutes with 100% CPU) and provides its results in a JSON format, which basically says for each recognized symbol: its shape together with its bounds within the original image. And then Audiveris OMR uses this symbol-based information.

This unfortunately does not cover the _training mode_ whereby you could train the model on thousands of scores. This training was performed on some specialized hardware available at ZHAW.

> 2. If you use sheet music from the MuseScore website (such a project was mentioned in a wiki page), I suppose MusicXML will also often be available (right?) - this is analoguous to my situation where I'd like to be able to tell Audiveris "this is the verified outcome I wanted" - would it be useful for the training to have something generating "validated .omr files" from MusicXML?

Well, surely not out-of-the-box! MusicXML is good at exchanging data between music applications. I can't tell for MuseScore .mscz format, but Audiveris .omr format contains more information than what we can find in MusicXML, especially when it comes to precise location in the input image. I'm not saying it would be impossible to kind of "reverse engineer" from MuseScore or MusicXML, I simply guess it would require significant efforts without any feasibility guarantee.

> 3. If I at some point get started scanning & proofing music, how would you like training materials contributed? Pull requests against the omr-dataset-tools project? The scanned images + the .omr files after using Audiveris UI to proof stuff is of interest?

Just the .omr files are sufficient. Because the black & white image of every input sheet is stored within the .omr file itself.
I don't know yet how we'll organize this kind of uploading workflow in a concrete manner. This is something we'll have to investigate further.

**Reply by hallvors** [reply]
Thanks for your very helpful and detailed responses. I may find time to test a alpha build and will report bugs if I run into things that seem worth reporting.

For a project I'm working on I'm using the Opensheetmusicdisplay library to transform MusicXML into SVG. OSMD can do batch conversion to SVG now. Perhaps that's interesting. When rendered (for example in headless Chrome) we get a DOM that can be queried to find positions of items.. I don't know to what extent you have roundtrip tests already, but I suppose MusicXML -> rendering -> OMR -> comparing generated MusicXML with originals could also be an interesting testing approach. Perhaps you do this already?
Again thanks for responding! I will not trouble you with more questions right now but hopefully I can at some point contribute some bug reports, debugging and/or validated OMR files.
